test/Runtime/Inputs/custom_rr_abi_utilities.h

```diff
@@ -0,0 +1,111 @@
+// specialized entrypoints. That's the same as ALL_REGS, minus x0 (the standard
+// entrypoint covers that), x16/x17 (temporary registers used as linker glue),
+// and x29 (the link register).
+#define FUNCTION_REGS(macro, ...)                                              \
+  macro( 1, __VA_ARGS__)                                                       \
+  macro( 2, __VA_ARGS__)                                                       \
+  macro( 3, __VA_ARGS__)                                                       \
+  macro( 4, __VA_ARGS__)                                                       \
+  macro( 5, __VA_ARGS__)                                                       \
+  macro( 6, __VA_ARGS__)                                                       \
+  macro( 7, __VA_ARGS__)                                                       \
+  macro( 8, __VA_ARGS__)                                                       \
+  macro( 9, __VA_ARGS__)                                                       \
+  macro(10, __VA_ARGS__)                                                       \
+  macro(11, __VA_ARGS__)                                                       \
+  macro(12, __VA_ARGS__)                                                       \
+  macro(13, __VA_ARGS__)                                                       \
+  macro(14, __VA_ARGS__)                                                       \
+  macro(15, __VA_ARGS__)                                                       \
+  macro(19, __VA_ARGS__)                                                       \
+  macro(20, __VA_ARGS__)                                                       \
+  macro(21, __VA_ARGS__)                                                       \
+  macro(22, __VA_ARGS__)                                                       \
+  macro(23, __VA_ARGS__)                                                       \
+  macro(24, __VA_ARGS__)                                                       \
+  macro(25, __VA_ARGS__)                                                       \
+  macro(26, __VA_ARGS__)                                                       \
+  macro(27, __VA_ARGS__)                                                       \
+  macro(28, __VA_ARGS__)
+
+// Apply `macro` to each function that gets specialized entrypoints. Also pass
+// 1 if the function is a retain variant, and 0 if it's a release variant.
+#define ALL_FUNCTIONS(macro)                                                   \
+  macro(swift_retain, 1)                                                       \
+  macro(swift_release, 0)                                                      \
+  macro(swift_bridgeObjectRetain, 1)                                           \
+  macro(swift_bridgeObjectRelease, 0)
+
+// Emit declarations for variables called xN stored in xN, initialized with
+// regs[N].
+#define PASS_REGS_HELPER(num) \
+  register void *x ## num asm ("x" #num) = regs[num];
```

Comment:
`register` is deprecated, we should avoid that.  Do we want to use `uintpt_t` instead of `void *`?

Reply:
You can't use `asm("<register>")` on a variable without also using `register`.

Reply:
As for the type, I don't think it matters whether you use `void *` or `uintptr_t` or, actually, `uint64_t` here, but `void *` is convenient because it lets us declare `regs` on the Swift side as `[UnsafeMutableRawPointer?]`, which then also lets us put object references into it without having to do an `unsafeBitCast()`.

Reply:
(I do _slightly_ worry that @compnerd might have a point here and that we should contemplate what happens if someone makes `register ... asm("<register>")` not work in the future. But in that case, the test will fail, so we'll notice that it's happened and can rewrite this code.)

Reply:
Right, we'll notice. Worst case we can rewrite this whole thing in inline assembly instead of convincing the compiler to load the registers for us.